[zhengyup] iP

zhengyup opened the pull request:
#Duke (HEAD edition)
> "SHOW ME WHAT YOU GOT"- Giant Head ([source](https://rickandmorty.fandom.com/wiki/Get_Schwifty_(episode)/Transcript))

DukePro frees your mind of having to remember things you need to do. It's,
- text based
- easy to learn 
- ~~FAST~~ SUPER FAST to use

All you need to do is, 
1. download it from [here](www.google.com)
2. double click it
3. add your tasks
4. let it manage your tasks for you 😎 

And it's **FREE**

Features : 
[x] Managing tasks
[  ] Managing deadlines (coming soon)
[  ] Reminders (coming soon)

-------

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
implemented skeletal version, changed duke to HEAD and changed logo

Commit:
completed level 1, echoes input and exits when input GET SCHWIFTY

Commit:
Level 2 completed, added an arraylist to keep track of inputs and prints the list when list is entered

Commit:
added class to represent Task and completed level 3

Commit:
created todo, event and deadline classes. yet to integrate to the functionality of Duke

Commit:
added functionality for level 4, works as expected from manual testing

Commit:
completed A-TEXTUiTesting, e

Commit:
added exception classes and handling

Commit:
implemented the deletion of tasks, and added a new error to handle selecting an index out of range for delete, mark and unmark operations. Not yet integrated

Commit:
added archive class to manage tasks. added abstraction to reduce the amount of code in Duke.java.

Commit:
Allow the saving and retrieval of tasks between uses of the chatbot

Users are not able to save tasks between separate uses of the chatbot.

using filereaders and writers we can save the tasks in a text file.

This gives a simple implementation of a local database

Commit:
Merge branch 'branch-Level-7'

Commit:
Add ability to understand date times given in dd/MM/yyyy HH:mm format and prompt for the correct format otherwise

Date was previously treated as a string, which creates problems when trying to do chronological operations on tasks.

Implemented using java LocalDateTime, receives in one format and prints in another

Commit:
Merge branch 'branch-Level-8'

Commit:
Refactor code to extract out closely related code as classes and as such use more OOP

Commit:
Merge branch 'A-MoreOOP'

Commit:
Restructure folders to place duke related files in the duke package

Commit:
Merge branch 'branch-A-Packages'

Commit:
Merge branch 'add-gradle-support' into branch-A-Gradle

Commit:
set up gradle for testing

Commit:
Merge branch 'branch-A-Gradle'

Commit:
add JUnit tests to test the loadTask function in storage and the markTask function in TaskList

Commit:
Merge branch 'branch-A-JUnit'

Commit:
add JavaDocs for all classes

Commit:
fix code violations

Commit:
add find functionality

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Resolve merge conflicts between the javadoc and coding standard

Commit:
Resolve merge conflict between master branch and branch-Level-9

Commit:
add checkstyle and fixed the errros pointed out by checkstyle

Commit:
Merge branch 'branch-A-Checkstyle'

Commit:
add gui capability and removed CLI UI

Commit:
Merge branch 'branch-Level-10'

Commit:
add step 4 for GUI integration

Tang-Moyan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,201 @@
+    public String getResponse(String input) {
+        String dukeProcessedText;
+        String[] parsedText = Parser.parse(input, tasks);
+        switch(parsedText[0]) {
+            case "mark" :
+                dukeProcessedText = tasks.markTask(Integer.parseInt(parsedText[1]));
+                break;
+            case "unmark" :
+                dukeProcessedText = tasks.unmarkTask(Integer.parseInt(parsedText[1]));
+                break;
+            case "delete" :
+                dukeProcessedText = tasks.deleteTask(Integer.parseInt(parsedText[1]));
+                break;
+            case "find" :
+                try {
+                    dukeProcessedText = tasks.find(parsedText[1]);
+                } catch (DukeException e) {
+                    dukeProcessedText = e.getMessage();
+                }
+                break;
+            case "list" :
+                dukeProcessedText = tasks.getAll();
+                break;
+            case "exception" :
+                dukeProcessedText = parsedText[1];
+                break;
+            case "todo" :
+            case "deadline" :
+            case "event" :
+                try {
+                    dukeProcessedText = tasks.addTask(parsedText);
+                } catch (DukeException e) {
+                    dukeProcessedText = e.getMessage();
+                }
+                break;
+            default:
+                dukeProcessedText = "Invalid input";
+                break;
```
There should be no indentation for case clauses as according to the Java coding standard. You can change the default arrangement of your IDE :).

Tang-Moyan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,92 @@
+package duke;
+
+import duke.exceptions.*;
```
According to the Java code standard, it is a good practice that imported classes should always be listed explicitly.

Tang-Moyan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,92 @@
+public class Parser {
+    public Parser(){}
+
+    public static String[] parse(String input, TaskList tasks) {
```
Do you think it would be more convenient for you as a developer and more robust for your programe to extract each token, separated by " " (space), first, and then process each token to decide whether it is a command, task name, integer or time? Something like that... 
I see what you are trying to do here. It is just that every time you want to add a new command, you have to modify this "if else" statement and check the length of the input and other things all over again.

Tang-Moyan commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,52 @@
+ * The UI class handles user interactions and printing messages in the Duke application.
+ */
+public class UI {
+    private final String logo = "       ___\n"
+            + "    . -^   `--,\n"
+            + "   /# =========`-_\n"
+            + "  /# (--====___====\\\n"
+            + " /#   .- --.  . --.|\n"
+            + "/##   |  * ) (   * ),\n"
+            + "|##   \\    /\\ \\   / |\n"
+            + "|###   ---   \\ ---  |\n"
+            + "|####      ___)    #|\n"
+            + "|######           ##|\n"
+            + " \\##### ---------- /\n"
+            + "  \\####           (\n"
+            + "   `\\###          |\n"
+            + "     \\###         |\n"
+            + "      \\##        |\n"
+            + "       \\###.    .)\n"
+            + "        `======/";
```
I like that you have your own logo! Very cool! :) 
ps: I am a Rick and Morty fan as well. Pickle Rick!

Tang-Moyan reviewed (commented):
Well done! I like your Rick logo. Generally good in coding quality. Variable names are appropriate.

nathanielcalimag commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Constructs a UI object.
+     */
+    public UI() {
```
As this constructor has no logic attached to it, should it be omitted?

nathanielcalimag commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,80 @@
+        return res;
+    }
+
+    /**
+     * Retrieves the start date of the event.
+     *
+     * @return The start date of the event.
+     */
+    public LocalDateTime getFrom() {
+        return this.from;
+    }
+
+    /**
+     * Retrieves the end date of the event.
+     *
+     * @return The end date of the event.
+     */
+    public LocalDateTime getTo() {
+        return to;
+    }
```
Should there be some consistency as to how you define these getter functions? (i.e. accessing instance fields through `this`)?

nathanielcalimag commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public class Main extends Application {
+
+    private Duke duke = new Duke();
+    @Override
+    public void start(Stage stage) {
```
Should there be a space between the `duke` field and the `start()` function?

nathanielcalimag commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,128 @@
+     *
+     * @param filePath The file path to store task data.
+     */
+
```
Is the space here between the comment and the function necessary?

nathanielcalimag commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,201 @@
+        userInput.clear();
+    }
+
+    public String getResponse(String input) {
```
I like the naming of your methods, ensuring that it is prefixed with a verb. This is in-line with the coding standards.

> Names representing methods must be verbs and written in camelCase.

Commit:
use assertions to check key assumptions are met

Commit:
rewrite code to adhere to code quality

Commit:
fix truncation

zhengyup committed:
Merge pull request #1 from zhengyup/branch-A-Assertions

use assertions to check key assumptions are met

Commit:
try solving merge commit 1

zhengyup committed:
Merge pull request #2 from zhengyup/branch-A-CodeQuality

Rewrite code to adhere to code quality standards

Commit:
add A-CI

Commit:
Add undo functionality to the chatbot

Commit:
Add undo functionality

Unable to undo changes made to the task list

Creating a stack of previous lists of tasks allows us to undo changes by simply setting the current list to the previous version by popping from the stack

This allows for a generic undo which is applicable undoing all commands

Commit:
set up website

Commit:
fix style errors using checkstyle and implement JUnit tests

Commit:
update tests to work with updated implementation

Commit:
furnish README as a user guide

Commit:
Merge branch 'A-UserGuide'
